src/modules/src/controller_indi.c

```diff
@@ -0,0 +1,311 @@
+                                         const uint32_t tick)
+{
+
+	/*
+	 * 1 - Update the gyro filter with the new measurements.
+	 */
+
+	float stateAttitudeRateRoll = radians(sensors->gyro.x);
+	float stateAttitudeRatePitch = -radians(sensors->gyro.y);
+	float stateAttitudeRateYaw = radians(sensors->gyro.z);
+
+	 struct FloatRates body_rates = {
+			  .p = stateAttitudeRateRoll,
+			  .q = stateAttitudeRatePitch,
+			  .r = stateAttitudeRateYaw,
+	 };
+	 filter_pqr(indi.rate, &body_rates);
+
+
+	/*
+	 * 2 - Calculate the derivative with finite difference.
+	 */
+
+	 finite_difference_from_filter(indi.rate_d, indi.rate);
+
+
+	/*
+	 * 3 - same filter on the actuators (or control_t values), using the commands from the previous timestep.
+	 */
+	 filter_pqr(indi.u, &indi.u_act_dyn);
+
+
+	/*
+	 * 4 - Calculate the desired angular acceleration by:
+	 * 4.1 - Rate_reference = P * attitude_error, where attitude error can be calculated with your favorite
+	 * algorithm. You may even use a function that is already there, such as attitudeControllerCorrectAttitudePID(),
+	 * though this will be inaccurate for large attitude errors, but it will be ok for now.
+	 * 4.2 Angular_acceleration_reference = D * (rate_reference – rate_measurement)
+	 */
+	 float referenceAttitudeRateRoll = radians(setpoint->attitudeRate.roll);
+	 float referenceAttitudeRatePitch = -radians(setpoint->attitudeRate.pitch);
+	 float referenceAttitudeRateYaw = radians(setpoint->attitudeRate.yaw);
+
+	 indi.angular_accel_ref.p = indi.reference_acceleration.rate_p * (referenceAttitudeRateRoll - body_rates.p);
+	 indi.angular_accel_ref.q = indi.reference_acceleration.rate_q * (referenceAttitudeRatePitch - body_rates.q);
+	 indi.angular_accel_ref.r = indi.reference_acceleration.rate_r * (referenceAttitudeRateYaw - body_rates.r);
+
+	/*
+	 * 5. Update the For each axis: delta_command = 1/control_effectiveness * (angular_acceleration_reference – angular_acceleration)
+	 */
+
+	 //Increment in angular acceleration requires increment in control input
+	 //G1 is the control effectiveness. In the yaw axis, we need something additional: G2.
+	 //It takes care of the angular acceleration caused by the change in rotation rate of the propellers
+	 //(they have significant inertia, see the paper mentioned in the header for more explanation)
+	 indi.du.p = 1.0f / indi.g1.p * (indi.angular_accel_ref.p - indi.rate_d[0]);
+	 indi.du.q = 1.0f / indi.g1.q * (indi.angular_accel_ref.q - indi.rate_d[1]);
+	 indi.du.r = 1.0f / (indi.g1.r + indi.g2) * (indi.angular_accel_ref.r - indi.rate_d[2] + indi.g2 * indi.du.r);
+
+
+	/*
+	 * 6. Add delta_commands to commands and bound to allowable values
+	*/
+
+	 indi.u_in.p = indi.u[0].o[0] + indi.du.p;
+	 indi.u_in.q = indi.u[1].o[0] + indi.du.q;
+	 indi.u_in.r = indi.u[2].o[0] + indi.du.r;
+
+	  //bound the total control input
+	if(STABILIZATION_INDI_FULL_AUTHORITY){
+	  clamp(indi.u_in.p, -9600.0f, 9600.0f);
```

Comment:
Here, use the maximum and minimum allowable command as specified for Crazyflie.

Reply:
Could not find this specification, but used some values inspired in this Mellinger controller code:
- https://github.com/amreelab/crazyflie-firmware/blob/d1fff167ce73647be9e93d2196a001e1d27816c5/src/modules/src/controller_mellinger.c#L289